[feat] 유저 도메인, 컨트롤러, 서비스, 리포지토리 생성 및 테스트

JeongUijeong opened the pull request:
- User.java 수정 및 UserVO.java 생성
- UserController 생성
- 회원가입, 로그인, 레벨업 기능 구현
- UserRepository 생성
- index, main, join, login 페이지 생성
- 유저 회원가입, 로그인, 레벨업 기능 테스트

JeongUijeong committed:
[feat] : 유저 도메인, 컨트롤러, 서비스, 리포지토리 생성

- User.java 수정 및 UserVO.java 생성
- UserController 생성
- 회원가입, 로그인, 레벨업 기능 구현
- UserRepository 생성
- index, main, join, login 페이지 생성

JeongUijeong committed:
[test] : 유저 회원가입, 로그인, 레벨업 기능 테스트

Nine-JH commented on .gitignore:
```diff
@@ -5,6 +5,10 @@ build/
 
+### yml ###
+*.yml
+*.yaml
```
회의 시간때 얘기를 했었는데 현재 코드는 현재 위치의 `.yaml` 파일만 ignore 하는 것이므로 모든 디렉토리에 있는 yaml을 제외하는 `**/*.yaml`으로 변경하시면 좋을 것 같습니다!
https://git-scm.com/docs/gitignore 또는 `gitignore 패턴` 키워드로 검색해보세요

JeongUijeong replied:
네! 수정했습니다~😁

Nine-JH commented on src/main/java/com/swger/tddstudy/user/controller/UserController.java:
```diff
@@ -0,0 +1,49 @@
+
+@Controller
+@RequiredArgsConstructor
+@RequestMapping("user")
```
url mapping prefix로 `user`를 주셨네요 저도 적용해봐야겠어요 😃

JeongUijeong replied:
😀👍🏻

Nine-JH commented on src/main/java/com/swger/tddstudy/user/repository/UserRepository.java:
```diff
@@ -0,0 +1,12 @@
+
+import java.util.Optional;
+
+@Repository
```
`@Repository`는 지우셔도 될 것 같아요! `JpaRepository` 인터페이스의 부모 클래스인 `org.springframework.data.repository` 인터페이스에는 이미 스프링 빈으로 등록해주는 마킹 어노테이션 `@Indexed` 가 존재한답니다

<br>
 
<img width="923" alt="스크린샷 2023-07-29 오후 7 19 48" src="https://github.com/fcstudy-project/tdd-study/assets/139187207/630f03be-446c-42a2-9826-f0e33d5b2478">

https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/stereotype/Indexed.html

JeongUijeong replied:
네! 이미지까지 올려주셔서 더 이해가 잘 됐습니다. 감사합니다!

Nine-JH commented on src/main/java/com/swger/tddstudy/user/domain/UserVO.java:
```diff
@@ -0,0 +1,33 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
```
`@Builder` 에너테이션을 작성하셨는데 따로 Builder 패턴을 사용한 코드는 없어 보이네요!
위의 toUserVO()에 한번 적용해보는건 어떠신가요?

JeongUijeong replied:
적용해봤습니다! 코드가 더 깔끔해진 것 같아요. 감사합니다~😀

Nine-JH commented on src/test/java/com/swger/tddstudy/user/UserTest.java:
```diff
@@ -0,0 +1,74 @@
+    @Rollback(value = true)
+    @DisplayName("로그인 테스트")
+    public void userloginTest() {
+        String username = "로그인용 이름";
```
위에서 생성하신 newUser() 메서드를 재활용할 수 있을 것 같은데 어떻게 생각하세요?

JeongUijeong replied:
그렇네요! 덕분에 불필요한 코드를 정리할 수 있었습니다 :D

Nine-JH commented on src/test/java/com/swger/tddstudy/user/UserTest.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @Test
+    @Transactional // test에서는 자동 롤백되긴 함
```
모든 메서드가 `@Transactional`이 적용된다면 클래스 레벨에서 `@Transactional`을 작성하셔도 좋을 것 같아요!

JeongUijeong replied:
좋은 의견 감사합니다! 반영했어요 😀

Nine-JH commented on src/test/java/com/swger/tddstudy/user/UserTest.java:
```diff
@@ -0,0 +1,74 @@
+        UserVO userVO = new UserVO(username, password, nickname, userLevel, userType);
+        Long savedId = userService.save(userVO);
+        UserVO levelUpResult = userService.levelUp(savedId);
+        // 레벨업 결과 BRONZE -> SILVER / SILVER -> GOLD / GOLD -> GOLD 면 성공
```
놓치기 쉬운 엣지 케이스인 GOLD -> GOLD 를 잘 테스트해주셨네요! 
일반적인 테스트케이스 (BRONZE -> SILVER or SILVER -> GOLD) 역시 추가해주면 좋을 것 같아요😀

JeongUijeong replied:
추가했습니다! 다른 테스트 코드들도 좀 더 꼼꼼하게 짜야겠네요 😂

Nine-JH commented on src/main/java/com/swger/tddstudy/user/domain/User.java:
```diff
@@ -1,31 +1,49 @@
     private Long id;
 
+    @NonNull
+    @Column(name = "username", nullable = false, length = 15)
```
DB 제약조건까지 잘 설정을 하셨네요!

JeongUijeong replied:
감사합니다!

Nine-JH commented on src/main/java/com/swger/tddstudy/user/controller/UserController.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @PostMapping("/login")
+    public String login(@ModelAttribute UserVO userVO, HttpSession session) {
```
로그인 분기처리와 세션 사용을 잘 해주셨네요!

JeongUijeong replied:
감사합니다! 😁

Nine-JH commented on src/main/java/com/swger/tddstudy/user/service/UserService.java:
```diff
@@ -0,0 +1,70 @@
+    private final UserRepository userRepository;
+
+    public Long save(UserVO userVO) {
+        return userRepository.save(userVO.toEntity()).getId();
```
UserVO에 대한 검증도 있었으면 좋겠어요!
* Entity에 제약조건이 있지만 이는 DBConnection 까지 얻은 뒤에 검증을 하게됩니다. 이전에 검증을 한다면 이 리소스를 아낄 수 있지 않을까요?
* Controller에서 VO에 관한 검증을 수행 해 Service의 역할을 줄이는 것도 좋아보여요

JeongUijeong replied:
네! 좋은 피드백 감사합니다~ 반영해보도록 하겠습니다!

Nine-JH reviewed (commented):
1등으로 제출을 해주셨네요 수고하셨습니다~!
HTML View 까지 구현해주신게 인상깊네요 👍
몇몇 코멘트를 작성했는데 확인부탁드려요~ 궁금하신점은 댓글 달아주세요
Comment에 👀 이모티콘을 남기시면 Resolve 처리 하겠습니다!

JeongUijeong committed:
[chore] : gitignore 수정 및 Order.java 오타 수정

JeongUijeong committed:
[feat] : RestController 생성 전 UserService 수정

- UserService의 save, login 메서드 반환값을 UserVO로 수정
- 컨트롤러 및 테스트 코드 내부 UserService 메서드 반환값 수정 적용

JeongUijeong committed:
[feat] : UserRestController 생성

- 회원가입, 로그인 api 구현

JeongUijeong committed:
[test] : 유저 컨트롤러(REST) 테스트 생성

JeongUijeong commented:
코멘트 감사합니다! 주신 피드백 적용해서 이후 코드 수정도 해보겠습니다~😀

JeongUijeong committed:
[feat] : PR 피드백을 반영한 코드 수정

- 클래스에 트랜잭션 애노테이션 사용
- UserRepository에 불필요한 @Repository 애노테이션 제거
- User 클래스에서 toUserVO 메서드에 빌더 사용

JeongUijeong committed:
[test] : newUser 메서드 사용 수정 및 레벨 업 테스트 세분화

JeongUijeong committed:
[feat] : 회원가입 양식 검사 추가

JeongUijeong committed:
[test] : 회원가입 및 로그인 실패 테스트 추가

ghrltjdtprbs commented:
제가 작성하던 코드보다 훨씬 간결하고 가독성이 좋네요! 저도 의정님 방식으로 코드를 다시 짜볼까 하는데 회의 시간에 간단한 코드 리뷰 부탁 드려도 될까요? 😂

ghrltjdtprbs reviewed (commented):
컨트롤러까지 꼼꼼하게 테스트 코드를 작성하신 점이 인상깊습니다!👍

KwonJuHwan commented:
mock에서 .exist로 NotNull을 쉽게 판단할 수 있다는 것을 배웠습니다. 간결한 코드가 보기 좋네요.

JeongUijeong committed:
[feat] : 유저 레벨 업 기능 수정

JeongUijeong committed:
[test] : 레벨 업 테스트 수정

JeongUijeong committed:
[feat] : Validation을 이용한 양식 검사를 수행하도록 수정

JeongUijeong committed:
[test] : Validation 기능 추가를 반영한 test 수정

ghrltjdtprbs reviewed (approved):
👍

kjyyjk reviewed (approved):
고생하셨습니당~
